src/server/node_func.c

```diff
@@ -2001,8 +2001,16 @@ fix_indirectness(resource *presc, struct pbsnode *pnode, int doit)
 				free_str(&ptargetan->rs_value);
 				ptargetan->rs_value.at_flags &= ~ATR_VFLAG_INDIRECT;
 			}
-
 			run_safety_check = 1;
+		} else {
+			/* Rescource assigned is not getting freed when a resource is set on an indirect rescource*/
+			ptargetan = find_resc_entry(&pnode->nd_attr[(int)ND_ATR_ResourceAssn], prdef);
+			if (consumable && (ptargetan != NULL) && (ptargetan->rs_value.at_flags & ATR_VFLAG_INDIRECT)) {
+				free_str(&ptargetan->rs_value);
+				prdef->rs_decode(&ptargetan->rs_value, NULL, NULL, NULL);
```

Comment:
What happens if there are running jobs on this vnode? If you just set it to 0 then the indirect resources assigned to the job on this vnode will suddenly become 0 and resources_assigned on the vnode from where the resource was being shared will never get decremented.

Reply:
You are right. The resources consumed by job are not getting released.
By looking more into the code, we are expecting old values in `porignalrc` and new values in `presc`. But `porignalrc` got modified before arriving at action function and it maintains the same values as `presc`. So any check in this function related to "porignalrc" simply doesnt work. One such check is the presence of job in the node we are setting the indirect resource. Another, check which is failing is when you try to set an indirect resource on a target resource as it would create a graph like structure. Then the third usage of `porignalrc` is to do what I fixed by adding another if condition.
An obvious way to fix all the three issues by checking `rescource_assigned` value for flags instead of `resource_available`. `resource_assigned` will still maintain the flags as it is not overwritten in `mgr_set_attr2`. I need to check the implications of fixing it this way.
These three cases also needs to be covered using automated test cases. I'll update you once the changes are ready.

Reply:
I don't think we should use resources_assigned to do the job. That just seems like a hack.
I'd suggest you create a setter function that you add to node's resources_available attribute and let that setter first call fix_indirectness and see whether we should even be setting the attribute. If it fails, don't set anything.

Reply:
The setter function is common for all resources - `set_resc`. It is a very generic function and it does not pass the parent object during set.

Reply:
Well I'm not asking you to change set_resc but to implement a function that will call set_resc inside it. I didn't notice the parent object, is there some other way to get that? I'll look into it tomorrow.

Reply:
Me and Bhroam had a long conversation about this issue, we both are in agreement that for short term using resources_assigned would solve the problem. Currently action functions are serving dual purpose (to validate the inputs and perform action). Unless we separate them out it is really hard to solve this problem. IMO validation should occur before we set anything but the reason it is done this way is to make sure that conflicting attributes do not get set at the same time (like walltime and soft_walltime).